How to deal with associated type families

Let's say I have a class with an associated type family:

```haskell
class IsList l where
  type Item l
  fromList :: [Item l] -> l
  toList   :: l -> [Item l]
```

Trying to feed this to `singletons` fails pretty early:

```
Bug.hs:13:3: error:
    Only method declarations can be promoted within a class.
   |
13 | $(singletons [d|
   |   ^^^^^^^^^^^^^^...
```

But need it fail? There appears to be a sensible way to promote and single `IsList`:

```haskell
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
module IsList where

import Data.Kind
import Data.List.NonEmpty (NonEmpty(..))
import Data.Singletons.Prelude.List.NonEmpty (Sing(..))
import Data.Singletons.Prelude

-----
-- Original class
-----

class IsList l where
  type Item l
  fromList :: [Item l] -> l
  toList   :: l -> [Item l]

-----
-- Promoted/singled version
-----

class PIsList (l :: Type) where
  type FromList (a :: [Item l]) :: l
  type ToList   (a :: l)        :: [Item l]

class SIsList l where
  sFromList :: forall (a :: [Item l]). Sing a -> Sing (FromList a :: l)
  sToList   :: forall (a :: l).        Sing a -> Sing (ToList a)

-----
-- Example instance
-----

instance IsList (NonEmpty a) where
  type Item (NonEmpty a) = a
  fromList (x:xs)   = x :| xs
  fromList []       = error "empty list"
  toList ~(x :| xs) = x:xs

instance PIsList (NonEmpty a) where
  type FromList (x:xs)  = x :| xs
  type FromList '[]     = Error "empty list"
  type ToList (x :| xs) = x:xs

instance SIsList (NonEmpty a) where
  sFromList (SCons x xs) = x :%| xs
  sFromList SNil         = sError (sing :: Sing "empty list")
  sToList (x :%| xs)     = SCons x xs
```

That is, promoting and singling leaves `Item` unaffected. This should work similarly for top-level type families as well.

Darn, I knew it wouldn't be that easy. If you try to promote this, you'll run into problems with type families pretty quickly. Here's a subset of generated code which illustrates the issue:

```haskell
{-# LANGUAGE ExistentialQuantification #-}                                                
{-# LANGUAGE ScopedTypeVariables #-}                                                      
{-# LANGUAGE TypeApplications #-}                                                         
{-# LANGUAGE TypeFamilies #-}                                                             
{-# LANGUAGE TypeInType #-}                                                               
{-# LANGUAGE UndecidableInstances #-}                                                     
module Bug where                                                                          
                                                                                          
import Data.Kind                                                                          
import Data.Singletons                                                                    
                                                                                          
class IsList (l :: Type) where                                                            
  type Item l                                                                             
  toList :: l -> [Item l]                                                                 
                                                                                          
type ToListSym1 (x :: l) = ToList x                                                       
                                                                                          
data ToListSym0 (x :: TyFun l [Item l])                                                   
  = forall arg. SameKind (Apply ToListSym0 arg) (ToListSym1 arg) =>
    ToListSym0KindInference

type instance Apply ToListSym0 x = ToList x

class PIsList (l :: Type) where
  type ToList (arg :: l) :: [Item l]
```
```
GHCi, version 8.2.0.20170523: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/rgscott/.ghci
[1 of 1] Compiling Bug              ( Bug.hs, interpreted )

Bug.hs:22:15: error:
    • Illegal type synonym family application in instance: [Item l]
    • In the type instance declaration for ‘Apply’
   |
22 | type instance Apply ToListSym0 x = ToList x
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

If you squint past that vague error message, you'll come to the horrifying realization that matching on `ToListSym0` also implicitly matches on its kind, which mentions `[Item l]`. Matching on other type families within a type family equation is strictly forbidden, alas.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -19,6 +19,11 @@ Changelog for singletons project
   allowing it to work over more open kinds. Also rename `SomeTypeRepStar` to
   `SomeTypeRepTYPE`, and change its definition accordingly.
 
+* Promoting or singling a type synonym or type family declaration now produces
+  defunctionalization symbols for it. (Previously, promoting or singling a type
+  synonym did nothing whatsoever, and promoting or singling a type family
+  produced an error.)
+
 * Add `(%<=?)`, a singled version of `(<=?)` from `GHC.TypeNats`, as well as
   defunctionalization symbols for `(<=?)`, to `Data.Singletons.TypeLits`.
 

--- a/README.md
+++ b/README.md
@@ -689,3 +689,6 @@ Known bugs
   problem is that a use of an associated type family tied to a class with
   fundeps doesn't provoke the fundep to kick in. This is GHC's problem, in
   the end.
+* Singled code that contains uses type families is likely to fail due to GHC
+  Trac #12564. Note that singling type family declarations themselves is fine
+  (and often desired, since that produces defunctionalization symbols for them).

--- a/src/Data/Singletons/Partition.hs
+++ b/src/Data/Singletons/Partition.hs
@@ -38,16 +38,20 @@ data PartitionedDecs =
         , pd_class_decs :: [UClassDecl]
         , pd_instance_decs :: [UInstDecl]
         , pd_data_decs :: [DataDecl]
+        , pd_ty_syn_decs :: [TySynDecl]
+        , pd_open_type_family_decs :: [OpenTypeFamilyDecl]
+        , pd_closed_type_family_decs :: [ClosedTypeFamilyDecl]
         , pd_derived_eq_decs :: [DerivedEqDecl]
         , pd_derived_show_decs :: [DerivedShowDecl]
         }
 
 instance Semigroup PartitionedDecs where
-  PDecs a1 b1 c1 d1 e1 f1 <> PDecs a2 b2 c2 d2 e2 f2 =
+  PDecs a1 b1 c1 d1 e1 f1 g1 h1 i1 <> PDecs a2 b2 c2 d2 e2 f2 g2 h2 i2 =
     PDecs (a1 <> a2) (b1 <> b2) (c1 <> c2) (d1 <> d2) (e1 <> e2) (f1 <> f2)
+          (g1 <> g2) (h1 <> h2) (i1 <> i2)
 
 instance Monoid PartitionedDecs where
-  mempty = PDecs [] [] [] [] [] []
+  mempty = PDecs [] [] [] [] [] [] [] [] []
   mappend = (<>)
 
 -- | Split up a @[DDec]@ into its pieces, extracting 'Ord' instances
@@ -76,12 +80,13 @@ partitionDec (DDataD nd _cxt name tvbs mk cons derivings) = do
       map (\p -> (strat, predToType p)) preds
 
 partitionDec (DClassD cxt name tvbs fds decs) = do
-  env <- concatMapM partitionClassDec decs
+  (lde, otfs) <- concatMapM partitionClassDec decs
   return $ mempty { pd_class_decs = [ClassDecl { cd_cxt       = cxt
                                                , cd_name      = name
                                                , cd_tvbs      = tvbs
                                                , cd_fds       = fds
-                                               , cd_lde       = env }] }
+                                               , cd_lde       = lde }]
+                  , pd_open_type_family_decs = otfs }
 partitionDec (DInstanceD _ cxt ty decs) = do
   defns <- liftM catMaybes $ mapM partitionInstanceDec decs
   (name, tys) <- split_app_tys [] ty
@@ -95,9 +100,18 @@ partitionDec (DInstanceD _ cxt ty decs) = do
     split_app_tys acc (DSigT t _)   = split_app_tys acc t
     split_app_tys _ _ = fail $ "Illegal instance head: " ++ show ty
 partitionDec (DRoleAnnotD {}) = return mempty  -- ignore these
-partitionDec (DTySynD {})     = return mempty  -- ignore type synonyms;
-                                               -- promotion is a no-op, and
-                                               -- singling expands all syns
+partitionDec (DTySynD name tvbs _type) =
+  -- See Note [Partitioning, type synonyms, and type families]
+  pure $ mempty { pd_ty_syn_decs = [TySynDecl name tvbs] }
+partitionDec (DClosedTypeFamilyD tf_head _) =
+  -- See Note [Partitioning, type synonyms, and type families]
+  pure $ mempty { pd_closed_type_family_decs = [TypeFamilyDecl tf_head] }
+partitionDec (DOpenTypeFamilyD tf_head) =
+  -- See Note [Partitioning, type synonyms, and type families]
+  pure $ mempty { pd_open_type_family_decs = [TypeFamilyDecl tf_head] }
+partitionDec (DTySynInstD {}) = pure mempty
+  -- There's no need to track type family instances, since
+  -- we already record the type family itself separately.
 partitionDec (DStandaloneDerivD mb_strat ctxt ty) =
   case unfoldType ty of
     cls_pred_ty :| cls_tys
@@ -122,15 +136,24 @@ partitionDec (DStandaloneDerivD mb_strat ctxt ty) =
 partitionDec dec =
   fail $ "Declaration cannot be promoted: " ++ pprint (decToTH dec)
 
-partitionClassDec :: Monad m => DDec -> m ULetDecEnv
-partitionClassDec (DLetDec (DSigD name ty)) = return $ typeBinding name ty
+partitionClassDec :: Monad m => DDec -> m (ULetDecEnv, [OpenTypeFamilyDecl])
+partitionClassDec (DLetDec (DSigD name ty)) =
+  pure (typeBinding name ty, mempty)
 partitionClassDec (DLetDec (DValD (DVarPa name) exp)) =
-  return $ valueBinding name (UValue exp)
+  pure (valueBinding name (UValue exp), mempty)
 partitionClassDec (DLetDec (DFunD name clauses)) =
-  return $ valueBinding name (UFunction clauses)
+  pure (valueBinding name (UFunction clauses), mempty)
 partitionClassDec (DLetDec (DInfixD fixity name)) =
-  return $ infixDecl fixity name
-partitionClassDec (DLetDec (DPragmaD {})) = return mempty
+  pure (infixDecl fixity name, mempty)
+partitionClassDec (DLetDec (DPragmaD {})) =
+  pure (mempty, mempty)
+partitionClassDec (DOpenTypeFamilyD tf_head) =
+  -- See Note [Partitioning, type synonyms, and type families]
+  pure (mempty, [TypeFamilyDecl tf_head])
+partitionClassDec (DTySynInstD {}) =
+  -- There's no need to track associated type family default equations, since
+  -- we already record the type family itself separately.
+  pure (mempty, mempty)
 partitionClassDec _ =
   fail "Only method declarations can be promoted within a class."
 
@@ -140,6 +163,9 @@ partitionInstanceDec (DLetDec (DValD (DVarPa name) exp)) =
 partitionInstanceDec (DLetDec (DFunD name clauses)) =
   return $ Just (name, UFunction clauses)
 partitionInstanceDec (DLetDec (DPragmaD {})) = return Nothing
+partitionInstanceDec (DTySynInstD {}) = pure Nothing
+  -- There's no need to track associated type family instances, since
+  -- we already record the type family itself separately.
 partitionInstanceDec _ =
   fail "Only method bodies can be promoted within an instance."
 
@@ -290,3 +316,18 @@ isNonVanillaDataType data_ty = anyM $ \con@(DCon _ ctxt _ _ _) -> do
     anyM p (x:xs) = do
       b <- p x
       if b then return True else anyM p xs
+
+{-
+Note [Partitioning, type synonyms, and type families]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The process of singling does not produce any new declarations corresponding to
+type synonyms or type families, so they are "ignored" in a sense. Nevertheless,
+we explicitly track them during partitioning, since we want to create
+defunctionalization symbols for them.
+
+Also note that:
+
+1. Other uses of type synonyms in singled code will be expanded away.
+2. Other uses of type families in singled code are unlikely to work at present
+   due to Trac #12564.
+-}

--- a/src/Data/Singletons/Promote.hs
+++ b/src/Data/Singletons/Promote.hs
@@ -188,12 +188,16 @@ promoteDecs :: [DDec] -> PrM ()
 promoteDecs raw_decls = do
   decls <- expand raw_decls     -- expand type synonyms
   checkForRepInDecls decls
-  PDecs { pd_let_decs          = let_decs
-        , pd_class_decs        = classes
-        , pd_instance_decs     = insts
-        , pd_data_decs         = datas
-        , pd_derived_eq_decs   = derived_eq_decs } <- partitionDecs decls
+  PDecs { pd_let_decs                = let_decs
+        , pd_class_decs              = classes
+        , pd_instance_decs           = insts
+        , pd_data_decs               = datas
+        , pd_ty_syn_decs             = ty_syns
+        , pd_open_type_family_decs   = o_tyfams
+        , pd_closed_type_family_decs = c_tyfams
+        , pd_derived_eq_decs         = derived_eq_decs } <- partitionDecs decls
 
+  defunTypeDecls ty_syns c_tyfams o_tyfams
     -- promoteLetDecs returns LetBinds, which we don't need at top level
   _ <- promoteLetDecs noPrefix let_decs
   mapM_ promoteClassDec classes

--- a/src/Data/Singletons/Promote/Defun.hs
+++ b/src/Data/Singletons/Promote/Defun.hs
@@ -15,6 +15,7 @@ import Data.Singletons.Promote.Monad
 import Data.Singletons.Promote.Type
 import Data.Singletons.Names
 import Language.Haskell.TH.Syntax
+import Data.Singletons.Syntax
 import Data.Singletons.Util
 import Control.Monad
 
@@ -30,26 +31,55 @@ defunInfo (DTyVarI _name _ty) =
 defunInfo (DPatSynI {}) =
   fail "Building defunctionalization symbols of pattern synonyms not supported"
 
+defunTypeDecls :: [TySynDecl]
+               -> [ClosedTypeFamilyDecl]
+               -> [OpenTypeFamilyDecl]
+               -> PrM ()
+defunTypeDecls ty_syns c_tyfams o_tyfams = do
+  defun_ty_syns <-
+    concatMapM (\(TySynDecl name tvbs) -> buildDefunSymsTySynD name tvbs) ty_syns
+  defun_c_tyfams <-
+    concatMapM (buildDefunSymsClosedTypeFamilyD . getTypeFamilyDecl) c_tyfams
+  defun_o_tyfams <-
+    concatMapM (buildDefunSymsOpenTypeFamilyD . getTypeFamilyDecl) o_tyfams
+  emitDecs $ defun_ty_syns ++ defun_c_tyfams ++ defun_o_tyfams
+
 buildDefunSyms :: DDec -> PrM [DDec]
 buildDefunSyms (DDataD _new_or_data _cxt _tyName _tvbs _k ctors _derivings) =
   buildDefunSymsDataD ctors
-buildDefunSyms (DClosedTypeFamilyD (DTypeFamilyHead name tvbs result_sig _) _) = do
-  let arg_m_kinds = map extractTvbKind tvbs
-  defunctionalize name arg_m_kinds (resultSigToMaybeKind result_sig)
-buildDefunSyms (DOpenTypeFamilyD (DTypeFamilyHead name tvbs result_sig _)) = do
-  let arg_kinds = map (default_to_star . extractTvbKind) tvbs
-      res_kind  = default_to_star (resultSigToMaybeKind result_sig)
-      default_to_star Nothing  = Just $ DConT typeKindName
-      default_to_star (Just k) = Just k
-  defunctionalize name arg_kinds res_kind
-buildDefunSyms (DTySynD name tvbs _type) = do
-  let arg_m_kinds = map extractTvbKind tvbs
-  defunctionalize name arg_m_kinds Nothing
+buildDefunSyms (DClosedTypeFamilyD tf_head _) =
+  buildDefunSymsClosedTypeFamilyD tf_head
+buildDefunSyms (DOpenTypeFamilyD tf_head) =
+  buildDefunSymsOpenTypeFamilyD tf_head
+buildDefunSyms (DTySynD name tvbs _type) =
+  buildDefunSymsTySynD name tvbs
 buildDefunSyms (DClassD _cxt name tvbs _fundeps _members) = do
   let arg_m_kinds = map extractTvbKind tvbs
   defunctionalize name arg_m_kinds (Just (DConT constraintName))
 buildDefunSyms _ = fail $ "Defunctionalization symbols can only be built for " ++
                           "type families and data declarations"
+
+buildDefunSymsClosedTypeFamilyD :: DTypeFamilyHead -> PrM [DDec]
+buildDefunSymsClosedTypeFamilyD = buildDefunSymsTypeFamilyHead id
+
+buildDefunSymsOpenTypeFamilyD :: DTypeFamilyHead -> PrM [DDec]
+buildDefunSymsOpenTypeFamilyD = buildDefunSymsTypeFamilyHead default_to_star
+  where
+    default_to_star :: Maybe DKind -> Maybe DKind
+    default_to_star Nothing  = Just $ DConT typeKindName
+    default_to_star (Just k) = Just k
+
+buildDefunSymsTypeFamilyHead
+  :: (Maybe DKind -> Maybe DKind) -> DTypeFamilyHead -> PrM [DDec]
+buildDefunSymsTypeFamilyHead default_kind (DTypeFamilyHead name tvbs result_sig _) = do
+  let arg_kinds = map (default_kind . extractTvbKind) tvbs
+      res_kind  = default_kind (resultSigToMaybeKind result_sig)
+  defunctionalize name arg_kinds res_kind
+
+buildDefunSymsTySynD :: Name -> [DTyVarBndr] -> PrM [DDec]
+buildDefunSymsTySynD name tvbs = do
+  let arg_m_kinds = map extractTvbKind tvbs
+  defunctionalize name arg_m_kinds Nothing
 
 buildDefunSymsDataD :: [DCon] -> PrM [DDec]
 buildDefunSymsDataD ctors =

--- a/src/Data/Singletons/Single.hs
+++ b/src/Data/Singletons/Single.hs
@@ -20,6 +20,7 @@ import Data.Singletons.Deriving.Enum
 import Data.Singletons.Deriving.Show
 import Data.Singletons.Util
 import Data.Singletons.Promote
+import Data.Singletons.Promote.Defun
 import Data.Singletons.Promote.Monad ( promoteM )
 import Data.Singletons.Promote.Type
 import Data.Singletons.Names
@@ -239,14 +240,18 @@ singInfo (DPatSynI {}) =
 singTopLevelDecs :: DsMonad q => [Dec] -> [DDec] -> q [DDec]
 singTopLevelDecs locals raw_decls = withLocalDeclarations locals $ do
   decls <- expand raw_decls     -- expand type synonyms
-  PDecs { pd_let_decs          = letDecls
-        , pd_class_decs        = classes
-        , pd_instance_decs     = insts
-        , pd_data_decs         = datas
-        , pd_derived_eq_decs   = derivedEqDecs
-        , pd_derived_show_decs = derivedShowDecs } <- partitionDecs decls
+  PDecs { pd_let_decs                = letDecls
+        , pd_class_decs              = classes
+        , pd_instance_decs           = insts
+        , pd_data_decs               = datas
+        , pd_ty_syn_decs             = ty_syns
+        , pd_open_type_family_decs   = o_tyfams
+        , pd_closed_type_family_decs = c_tyfams
+        , pd_derived_eq_decs         = derivedEqDecs
+        , pd_derived_show_decs       = derivedShowDecs } <- partitionDecs decls
 
   ((letDecEnv, classes', insts'), promDecls) <- promoteM locals $ do
+    defunTypeDecls ty_syns c_tyfams o_tyfams
     promoteDataDecs datas
     (_, letDecEnv) <- promoteLetDecs noPrefix letDecls
     classes' <- mapM promoteClassDec classes

--- a/src/Data/Singletons/Syntax.hs
+++ b/src/Data/Singletons/Syntax.hs
@@ -45,8 +45,23 @@ instance Monoid PromDPatInfos where
 -- See @Note [Singling pattern signatures]@.
 type SingDSigPaInfos = [(DExp, DType)]
 
-  -- the relevant part of declarations
-data DataDecl      = DataDecl NewOrData Name [DTyVarBndr] [DCon] [DPred]
+-- The parts of data declarations that are relevant to singletons.
+data DataDecl = DataDecl NewOrData Name [DTyVarBndr] [DCon] [DPred]
+
+-- The parts of type synonyms that are relevant to singletons.
+data TySynDecl = TySynDecl Name [DTyVarBndr]
+
+-- The parts of open type families that are relevant to singletons.
+type OpenTypeFamilyDecl = TypeFamilyDecl 'Open
+
+-- The parts of closed type families that are relevant to singletons.
+type ClosedTypeFamilyDecl = TypeFamilyDecl 'Closed
+
+-- The parts of type families that are relevant to singletons.
+newtype TypeFamilyDecl (info :: FamilyInfo)
+  = TypeFamilyDecl { getTypeFamilyDecl :: DTypeFamilyHead }
+-- Whether a type family is open or closed.
+data FamilyInfo = Open | Closed
 
 data ClassDecl ann = ClassDecl { cd_cxt  :: DCxt
                                , cd_name :: Name

--- a/tests/SingletonsTestSuite.hs
+++ b/tests/SingletonsTestSuite.hs
@@ -90,6 +90,7 @@ tests =
     , compileAndDumpStdTest "TypeRepTYPE"
     , compileAndDumpStdTest "T297"
     , compileAndDumpStdTest "T312"
+    , compileAndDumpStdTest "T313"
     , compileAndDumpStdTest "T316"
     ],
     testCompileAndDumpGroup "Promote"

--- a/tests/compile-and-dump/Singletons/T178.ghc84.template
+++ b/tests/compile-and-dump/Singletons/T178.ghc84.template
@@ -14,6 +14,7 @@ Singletons/T178.hs:(0,0)-(0,0): Splicing declarations
     type U = [(Symbol, Occ)]
     empty :: U
     empty = []
+    type USym0 = U
     type StrSym0 = Str
     type OptSym0 = Opt
     type ManySym0 = Many

--- a/tests/compile-and-dump/Singletons/T313.ghc84.template
+++ b/tests/compile-and-dump/Singletons/T313.ghc84.template
@@ -1,0 +1,126 @@
+Singletons/T313.hs:(0,0)-(0,0): Splicing declarations
+    promote
+      [d| type PFoo1 a = Maybe a
+          type family PFoo2 a
+          type family PFoo3 a where
+            PFoo3 a = Maybe a
+          class PC (a :: Type) where
+            type PFoo4 a
+            type PFoo4 a = Maybe a
+          
+          type instance PFoo2 a = Maybe a
+          instance PC a where
+            type PFoo4 a = Maybe a |]
+  ======>
+    type PFoo1 a = Maybe a
+    type family PFoo2 a
+    type instance PFoo2 a = Maybe a
+    type family PFoo3 a where
+      PFoo3 a = Maybe a
+    class PC (a :: Type) where
+      type PFoo4 a
+      type PFoo4 a = Maybe a
+    instance PC a where
+      type PFoo4 a = Maybe a
+    type PFoo1Sym1 t = PFoo1 t
+    instance SuppressUnusedWarnings PFoo1Sym0 where
+      suppressUnusedWarnings
+        = snd ((GHC.Tuple.(,) PFoo1Sym0KindInference) GHC.Tuple.())
+    data PFoo1Sym0 l :: Type
+      where
+        PFoo1Sym0KindInference :: forall l arg.
+                                  SameKind (Apply PFoo1Sym0 arg) (PFoo1Sym1 arg) => PFoo1Sym0 l
+    type instance Apply PFoo1Sym0 l = PFoo1 l
+    type PFoo3Sym1 t = PFoo3 t
+    instance SuppressUnusedWarnings PFoo3Sym0 where
+      suppressUnusedWarnings
+        = snd ((GHC.Tuple.(,) PFoo3Sym0KindInference) GHC.Tuple.())
+    data PFoo3Sym0 l :: Type
+      where
+        PFoo3Sym0KindInference :: forall l arg.
+                                  SameKind (Apply PFoo3Sym0 arg) (PFoo3Sym1 arg) => PFoo3Sym0 l
+    type instance Apply PFoo3Sym0 l = PFoo3 l
+    type PFoo2Sym1 (t :: Type) = PFoo2 t
+    instance SuppressUnusedWarnings PFoo2Sym0 where
+      suppressUnusedWarnings
+        = snd ((GHC.Tuple.(,) PFoo2Sym0KindInference) GHC.Tuple.())
+    data PFoo2Sym0 (l :: TyFun Type Type) :: Type
+      where
+        PFoo2Sym0KindInference :: forall l arg.
+                                  SameKind (Apply PFoo2Sym0 arg) (PFoo2Sym1 arg) => PFoo2Sym0 l
+    type instance Apply PFoo2Sym0 l = PFoo2 l
+    type PFoo4Sym1 (t :: Type) = PFoo4 t
+    instance SuppressUnusedWarnings PFoo4Sym0 where
+      suppressUnusedWarnings
+        = snd ((GHC.Tuple.(,) PFoo4Sym0KindInference) GHC.Tuple.())
+    data PFoo4Sym0 (l :: TyFun Type Type) :: Type
+      where
+        PFoo4Sym0KindInference :: forall l arg.
+                                  SameKind (Apply PFoo4Sym0 arg) (PFoo4Sym1 arg) => PFoo4Sym0 l
+    type instance Apply PFoo4Sym0 l = PFoo4 l
+    class PPC (a :: Type)
+    instance PPC a
+Singletons/T313.hs:(0,0)-(0,0): Splicing declarations
+    singletons
+      [d| type SFoo1 a = Maybe a
+          type family SFoo2 a
+          type family SFoo3 a where
+            SFoo3 a = Maybe a
+          class SC (a :: Type) where
+            type SFoo4 a
+            type SFoo4 a = Maybe a
+          
+          type instance SFoo2 a = Maybe a
+          instance SC a where
+            type SFoo4 a = Maybe a |]
+  ======>
+    type SFoo1 a = Maybe a
+    type family SFoo2 a
+    type instance SFoo2 a = Maybe a
+    type family SFoo3 a where
+      SFoo3 a = Maybe a
+    class SC (a :: Type) where
+      type SFoo4 a
+      type SFoo4 a = Maybe a
+    instance SC a where
+      type SFoo4 a = Maybe a
+    type SFoo1Sym1 t = SFoo1 t
+    instance SuppressUnusedWarnings SFoo1Sym0 where
+      suppressUnusedWarnings
+        = snd ((GHC.Tuple.(,) SFoo1Sym0KindInference) GHC.Tuple.())
+    data SFoo1Sym0 l :: Type
+      where
+        SFoo1Sym0KindInference :: forall l arg.
+                                  SameKind (Apply SFoo1Sym0 arg) (SFoo1Sym1 arg) => SFoo1Sym0 l
+    type instance Apply SFoo1Sym0 l = SFoo1 l
+    type SFoo3Sym1 t = SFoo3 t
+    instance SuppressUnusedWarnings SFoo3Sym0 where
+      suppressUnusedWarnings
+        = snd ((GHC.Tuple.(,) SFoo3Sym0KindInference) GHC.Tuple.())
+    data SFoo3Sym0 l :: Type
+      where
+        SFoo3Sym0KindInference :: forall l arg.
+                                  SameKind (Apply SFoo3Sym0 arg) (SFoo3Sym1 arg) => SFoo3Sym0 l
+    type instance Apply SFoo3Sym0 l = SFoo3 l
+    type SFoo2Sym1 (t :: Type) = SFoo2 t
+    instance SuppressUnusedWarnings SFoo2Sym0 where
+      suppressUnusedWarnings
+        = snd ((GHC.Tuple.(,) SFoo2Sym0KindInference) GHC.Tuple.())
+    data SFoo2Sym0 (l :: TyFun Type Type) :: Type
+      where
+        SFoo2Sym0KindInference :: forall l arg.
+                                  SameKind (Apply SFoo2Sym0 arg) (SFoo2Sym1 arg) => SFoo2Sym0 l
+    type instance Apply SFoo2Sym0 l = SFoo2 l
+    type SFoo4Sym1 (t :: Type) = SFoo4 t
+    instance SuppressUnusedWarnings SFoo4Sym0 where
+      suppressUnusedWarnings
+        = snd ((GHC.Tuple.(,) SFoo4Sym0KindInference) GHC.Tuple.())
+    data SFoo4Sym0 (l :: TyFun Type Type) :: Type
+      where
+        SFoo4Sym0KindInference :: forall l arg.
+                                  SameKind (Apply SFoo4Sym0 arg) (SFoo4Sym1 arg) => SFoo4Sym0 l
+    type instance Apply SFoo4Sym0 l = SFoo4 l
+    class PSC (a :: Type)
+    instance PSC a
+    class SSC (a :: Type)
+    instance SSC a

--- a/tests/compile-and-dump/Singletons/T313.hs
+++ b/tests/compile-and-dump/Singletons/T313.hs
@@ -1,0 +1,38 @@
+module T313 where
+
+import Data.Kind
+import Data.Singletons.TH
+
+$(promote [d|
+  type PFoo1 a = Maybe a
+
+  type family   PFoo2 a
+  type instance PFoo2 a = Maybe a
+
+  type family PFoo3 a where
+    PFoo3 a = Maybe a
+
+  class PC (a :: Type) where
+    type PFoo4 a
+    type PFoo4 a = Maybe a
+
+  instance PC a where
+    type PFoo4 a = Maybe a
+  |])
+
+$(singletons [d|
+  type SFoo1 a = Maybe a
+
+  type family   SFoo2 a
+  type instance SFoo2 a = Maybe a
+
+  type family SFoo3 a where
+    SFoo3 a = Maybe a
+
+  class SC (a :: Type) where
+    type SFoo4 a
+    type SFoo4 a = Maybe a
+
+  instance SC a where
+    type SFoo4 a = Maybe a
+  |])

